### ISSUE TITLE
[sklearn#34006] RFC/META Harden CI security

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#362]
**Mirrored from [scikit-learn#34006](https://github.com/scikit-learn/scikit-learn/issues/34006)**

Original title: RFC/META Harden CI security
Original labels: 
Created: 2026-05-13T09:22:00Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

Because of the ongoing Mini Shai Hulud attack against CI & OSS supply chain (see [Postmortem: TanStack npm supply-chain compromise](https://tanstack.com/blog/npm-supply-chain-compromise-postmortem) for details), I think we should review and possibly straighten our CI before releasing 1.9.

In particular, we use the unsafe `pull_request_target` in several workflows which is a documented attack vector of the Mini Shai-Hulud worm that currently still spreads by corrupting an increasing number of npm and pypi packages.

## Short term:

- Regularly check that the dates of https://pypi.org/project/scikit-learn/#files match our manually triggered release dates (this seems to be the case as of today 2026-05-13)
- Review our `publish_pypi.yml` and `wheels.yml` workflows to make sure it is not subject by cache poisoning attacks (from a cursory glance it does not think we use any build cache, but we should double-check).
- Review findings reported by `zizmor .github/workflows/publish_pypi.yml` and `zizmor .github/workflows/wheels.yml`.
- Assess if workflows that use `pull_request_target` can have a side effect on our release process.
  - Note: I asked the following to Claude: https://claude.ai/share/c6aea06a-a9dc-4e55-b5f4-830254166994 (TL;DNR it isn't too worried, but we should probably spend more time reading this analysis in more details).
- Unzip and review the contents of the generated wheel files in the release branch prior to uploading to look for unexpected payload.
  - EDIT: I asked Claude (same conversation as the one linked above) to do that on the latest nightly wheel and it did not find anything fishy.

## Medium/Long term

- Hash-pin all github actions used in the `publish_pypi.yml` and `wheels.yml` workflows.
- Use relative dependency cooldown for GitHub Actions updates via dependabot:
  - https://docs.github.com/en/code-security/reference/supply-chain-security/dependabot-options-reference?versionId=free-pro-team%40latest&productId=code-security&restPage=tutorials%2Csecure-your-dependencies%2Cautomating-dependabot-with-github-actions#cooldown-
- Find ways to avoid the use of `pull_request_target` in our regular dev CI.
- Address any remaining "findings" reported by running `zizmor .` in our repo (currently `191 findings (73 suppressed, 32 fixable): 4 informational, 26 low, 1 medium, 87 high`, many of which are related to `pull_request_target` and unpinned action references).
- Use relative dependency cooldowns when generating lock files as soon as implemented in mamba and conda-lock

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [50](https://github.com/godofecht/flow-scikit/issues/50) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [51](https://github.com/godofecht/flow-scikit/issues/51) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
